feat: add live rates carriers list section

frosso opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Description
<!-- Explain the motivation for making this change -->

Adding a "Live rates at checkout" section on the settings page.
Some details are on the ticket.

Checked with UX - a few things that are ok for this first phase:
- the tooltip’s width cannot be modified and I also cannot add artificial line breaks - everything needs to go into one line: https://d.pr/i/I29c68
- I tried to keep the alignment of the columns consistent - but on small screens it becomes evident that we’re trying to squeeze a lot of content on a small column: https://d.pr/i/LwYhQB
  - on even smaller screens the whole app is a bit messed up

### Related issue(s)

<!--
  Is there an issue open this PR addresses? If so, link to it for more information.
  GitHub link example: "Fixes #1234" Syntax: `[close|closes|closed|fix|fixes|fixed|resolve|resolves|resolved] #1234`
  Note: Remove this section if this PR does not have a related issue.
-->

Closes https://github.com/Automattic/woocommerce-shipping-issues/issues/188

### Steps to reproduce & screenshots/GIFs

<!--
  Please include steps to reproduce, as well as screenshots or GIFs, showing the before & after.
  This helps expedite review and increase confidence for approval & merge.
-->

When testing against production (when testing against local/staging you might need to remove `DhlExpressAccount` from the `enabled_carrier_accounts` setting on the connect server).

Brand new site, with no flags:
![Screen Shot 2021-03-12 at 10 09 28 AM](https://user-images.githubusercontent.com/273592/110969998-ed64dd00-831e-11eb-868c-c1bd9eeb83e8.png)

With UPS beta:
![Screen Shot 2021-03-12 at 10 10 02 AM](https://user-images.githubusercontent.com/273592/110970047-f8b80880-831e-11eb-9891-4858bc39e7be.png)

Only grandfathered live rates:
![Screen Shot 2021-03-12 at 10 10 30 AM](https://user-images.githubusercontent.com/273592/110970092-03729d80-831f-11eb-9bf6-80dd6e2920de.png)


UPS beta, grandfathered live rates, DHL Express live rates:
![Screen Shot 2021-03-12 at 10 10 58 AM](https://user-images.githubusercontent.com/273592/110970146-108f8c80-831f-11eb-9afd-f158dc538732.png)


### Checklist

<!-- All testable code should have tests. -->
- [x] unit tests

<!-- An entry in the changelog file is always required -->
- [x] `changelog.txt` entry added

frosso commented on classes/class-wc-connect-settings-pages.php:
```diff
@@ -4,9 +4,9 @@
-		 * @array
+		 * @var WC_Connect_Service_Schemas_Store
 		 */
-		protected $fieldsets;
```
`fieldsets` is not used anywhere - cleaning up

frosso commented on classes/class-wc-connect-settings-pages.php:
```diff
@@ -80,10 +81,13 @@ public function output_shipping_settings_screen() {
+
 			$carriers_response = $this->api_client->get_carrier_accounts();
 			if ( ! is_wp_error( $carriers_response ) && $carriers_response ) {
-				$extra_args['carriers'] = $carriers_response->carriers;
```
Changed argument `carriers` to `carrier_accounts`, to generate less possible confusion when the live rates' "Carriers" are also displayed

frosso commented on client/apps/shipping-settings/index.js:
```diff
@@ -18,7 +18,7 @@ import { mergeHandlers } from 'state/action-watchers/utils';
 import { middleware as rawWpcomApiMiddleware } from 'state/data-layer/wpcom-api-middleware';
 import { combineReducers } from 'state/utils';
 
-export default ( { order_id: orderId, order_href: orderHref, carrier: carrier, continents, carriers, subscriptions } ) => ( {
```
No need to alias `carrier: carrier` - renamed `carriers` to `carrier_accounts: carrierAccounts`

frosso commented on client/apps/shipping-settings/index.js:
```diff
@@ -74,5 +74,5 @@ export default ( { order_id: orderId, order_href: orderHref, carrier: carrier, c
 		return [ rawWpcomApiMiddleware( mergeHandlers( wcsUiDataLayer, actionList ) ) ];
 	},
 
-	View: () => <ViewWrapper orderId={ orderId } orderHref={ orderHref } carrier={ carrier } carriers={ carriers } subscriptions = { subscriptions } />,
```
Looks like there was some extra space on `subscriptions = { ... }` - renamed `carriers` and added `liveRatesTypes`

frosso commented on client/extensions/woocommerce/woocommerce-services/views/carrier-accounts/list-item.js:
```diff
@@ -22,11 +22,11 @@ import { errorNotice as errorNoticeAction, successNotice as successNoticeAction
 
 const CarrierAccountListItem = ( props ) => {
-	const { data, translate, errorNotice, successNotice, siteId } = props;
+	const { accountData, translate, errorNotice, successNotice, siteId } = props;
```
Just renames on the prop

frosso commented on client/extensions/woocommerce/woocommerce-services/views/carrier-accounts/test/index.js:
```diff
@@ -38,7 +38,7 @@ describe( 'Carrier Accounts', () => {
 		const wrapper = mount(
 			<Wrapper>
-				<CarrierAccounts carriers={[
+				<CarrierAccounts accounts={[
```
Just renames on the prop

Ferdev replied:
:information_source: Missing spaces inside braces.

frosso commented on client/extensions/woocommerce/woocommerce-services/views/carrier-accounts/index.js:
```diff
@@ -15,8 +15,8 @@ import Card from 'components/card';
 
-const CarrierAccounts = ({ translate, carriers = [] }) => {
-	if ( carriers.length === 0 ) {
+const CarrierAccounts = ({ translate, accounts = [] }) => {
```
Just renames on the prop

Ferdev replied:
:information_source: Missing spaces inside parens.

frosso replied:
Doesn't seem to be enforced on `[]` anywhere else.

frosso commented on client/extensions/woocommerce/woocommerce-services/views/live-rates-carriers-list/index.js:
```diff
@@ -0,0 +1,31 @@
+import ExtendedHeader from 'woocommerce/components/extended-header'
+import CarriersList from './carriers-list'
+
+const supportedCarrierIds = ['wc_services_usps', 'wc_services_dhlexpress'];
```
When testing against staging, there's also FedEx - need to whitelist the carriers to generate less possible confusion when other people are testing against the staging server

Ferdev replied:
:information_source: Missing spaces inside brackets.

Ferdev reviewed (approved):
Looks good to me! 

The only problem I found is with code style: there are some places were there aren't spaces inside parenthesis and braces.

But the code looks good and works beautifully. Thanks!

Ferdev commented on client/extensions/woocommerce/woocommerce-services/views/carrier-accounts/list-item.js:
```diff
@@ -22,11 +22,11 @@ import { errorNotice as errorNoticeAction, successNotice as successNoticeAction
 	const [isDisconnectDialogVisible, setIsDisconnectDialogVisible] = React.useState(false);
 	const [isSaving, setIsSaving] = React.useState(false);
-	const [carrierId, setCarrierId] = React.useState(data.id);
+	const [carrierId, setCarrierId] = React.useState(accountData.id);
```
:information_source: Missing spaces inside parens and brackets.

frosso replied:
I didn't change the formatting of this file - looks like it can be fixed by https://github.com/Automattic/woocommerce-services/issues/2097

Ferdev commented on client/extensions/woocommerce/woocommerce-services/views/carrier-accounts/list-item.js:
```diff
@@ -88,13 +88,13 @@ const CarrierAccountListItem = ( props ) => {
 		<div className="carrier-accounts__list-item">
 			<div className="carrier-accounts__list-item-carrier-icon">
-				<CarrierIcon carrier={ carrierTypeIconMap[data.type] } size={ 18 } />
+				<CarrierIcon carrier={ carrierTypeIconMap[accountData.type] } size={ 18 } />
```
:information_source: Missing spaces inside brackets.

Ferdev commented on client/extensions/woocommerce/woocommerce-services/views/carrier-accounts/list-item.js:
```diff
@@ -104,7 +104,7 @@ const CarrierAccountListItem = ( props ) => {
 					<a
 						href={
-							`admin.php?page=wc-settings&tab=shipping&section=woocommerce-services-settings&carrier=${data.type}`
+							`admin.php?page=wc-settings&tab=shipping&section=woocommerce-services-settings&carrier=${accountData.type}`
```
:information_source: Missing spaces inside braces.

Ferdev commented on client/extensions/woocommerce/woocommerce-services/views/carrier-accounts/test/index.js:
```diff
@@ -71,7 +71,7 @@ describe( 'Carrier Accounts', () => {
 		const wrapper = mount(
 			<Wrapper>
-				<CarrierAccounts carriers={[
+				<CarrierAccounts accounts={[
```
:information_source: Missing spaces inside braces.

Ferdev commented on client/extensions/woocommerce/woocommerce-services/views/carrier-accounts/test/index.js:
```diff
@@ -103,7 +103,7 @@ describe( 'Carrier Accounts', () => {
 		const wrapper = mount(
 			<Wrapper>
-				<CarrierAccounts carriers={[
+				<CarrierAccounts accounts={[
```
:information_source: Missing spaces inside braces.

Ferdev commented on client/extensions/woocommerce/woocommerce-services/views/live-rates-carriers-list/carriers-list.js:
```diff
@@ -0,0 +1,119 @@
+import CarrierIcon from '../../components/carrier-icon'
+import Gridicon from 'gridicons'
+
+const Actions = localize(({ translate }) => {
```
:information_source: Missing spaces inside parens and braces.

Ferdev commented on client/extensions/woocommerce/woocommerce-services/views/live-rates-carriers-list/carriers-list.js:
```diff
@@ -0,0 +1,119 @@
+	return (
+		<>
+			{ /* eslint-disable-next-line wpcalypso/jsx-classname-namespace */}
+			<a className="button is-compact" href="admin.php?page=wc-settings&tab=shipping&section">{translate('Add to shipping zones')}</a>
```
:information_source: Missing spaces inside parens and braces.

Ferdev commented on client/extensions/woocommerce/woocommerce-services/views/live-rates-carriers-list/carriers-list.js:
```diff
@@ -0,0 +1,119 @@
+			<a className="button is-compact" href="admin.php?page=wc-settings&tab=shipping&section">{translate('Add to shipping zones')}</a>
+			<Tooltip
+				position="top left"
+				text={translate('To be displayed at checkout, this carrier must be added as a shipping method to selected shipping zones')}
```
:information_source: Missing spaces inside parens and braces.

Ferdev commented on client/extensions/woocommerce/woocommerce-services/views/live-rates-carriers-list/carriers-list.js:
```diff
@@ -0,0 +1,119 @@
+				text={translate('To be displayed at checkout, this carrier must be added as a shipping method to selected shipping zones')}
+			>
+				<div>
+					<Gridicon icon="help-outline" size={18}/>
```
:information_source: Missing spaces inside braces.

Ferdev commented on client/extensions/woocommerce/woocommerce-services/views/live-rates-carriers-list/carriers-list.js:
```diff
@@ -0,0 +1,119 @@
+	translate,
+	name,
+	amount
+}) => translate('Up to {{strong}}%(discountAmount)s{{/strong}} off on %(carrierName)s shipping labels', {
```
:information_source: Missing spaces inside parens and braces.

Ferdev commented on client/extensions/woocommerce/woocommerce-services/views/live-rates-carriers-list/carriers-list.js:
```diff
@@ -0,0 +1,119 @@
+}))
+
+const carrierItemMap = {
+	'wc_services_usps': ({ translate }) => (
```
:information_source: Missing spaces inside parens/braces.

Ferdev commented on client/extensions/woocommerce/woocommerce-services/views/live-rates-carriers-list/carriers-list.js:
```diff
@@ -0,0 +1,119 @@
+			<div className="live-rates-carriers-list__carrier">{translate('USPS')}</div>
+			<div className="live-rates-carriers-list__features">
+				<ul>
+					<li>{translate('Ship with the largest delivery network in the United States')}</li>
```
:information_source: Missing spaces inside parens/braces.

Ferdev commented on client/extensions/woocommerce/woocommerce-services/views/live-rates-carriers-list/carriers-list.js:
```diff
@@ -0,0 +1,119 @@
+						<CarrierDiscount name={translate('USPS')} amount="81%"/>
+					</li>
+					<li>
+						{translate('Live rates for %(carrierName)s at checkout', {
+							args: {
+								carrierName: translate('USPS'),
+							},
+						})}
```
:information_source: Missing spaces inside parens/braces.

Ferdev commented on client/extensions/woocommerce/woocommerce-services/views/live-rates-carriers-list/carriers-list.js:
```diff
@@ -0,0 +1,119 @@
+	),
+}
+
+const CarriersList = ({ translate, carrierIds }) => {
+	return (
+		<Card className="live-rates-carriers-list__wrapper">
+			<div className="live-rates-carriers-list__heading">
+				<div className="live-rates-carriers-list__icon"/>
+				<div className="live-rates-carriers-list__carrier">{translate('Carrier')}</div>
+				<div className="live-rates-carriers-list__features">{translate('Features')}</div>
+				<div className="live-rates-carriers-list__actions"/>
+			</div>
+			{carrierIds.map(carrierId => {
+				const CarrierView = carrierItemMap[carrierId]
+				if (!CarrierView) {
+					return null
+				}
+
+				return (
+					<CarrierView key={carrierId} translate={translate}/>
+				)
+			})}
+		</Card>
+	)
+}
+
+export default localize(CarriersList)
```
:information_source: Missing spaces inside parens/braces.

Ferdev commented on client/extensions/woocommerce/woocommerce-services/views/live-rates-carriers-list/index.js:
```diff
@@ -0,0 +1,31 @@
+
+const supportedCarrierIds = ['wc_services_usps', 'wc_services_dhlexpress'];
+
+const LiveRatesCarriersList = ({ translate, carrierIds }) => {
+	if (carrierIds.some(carrierId => supportedCarrierIds.includes(carrierId)) === false) {
+		return null
+	}
+
+	return (
+		<div>
+			<ExtendedHeader
+				label={translate('Live rates at checkout')}
+				description={translate('Show live rates directly on your store - never under or overcharge for shipping again')}
+			/>
+			<CarriersList carrierIds={carrierIds} />
+		</div>
+	)
+}
+
+export default localize(LiveRatesCarriersList)
```
:information_source: Missing spaces inside parens/braces.

Ferdev commented on client/extensions/woocommerce/woocommerce-services/views/live-rates-carriers-list/test/live-rates-carriers-list.test.js:
```diff
@@ -0,0 +1,36 @@
+import LiveRatesCarriersList from '..'
+
+/* eslint-disable wpcalypso/jsx-classname-namespace */
+describe('LiveRatesCarriersList', () => {
+	it('should not render when the list of accounts is empty', () => {
+		const wrapper = mount(<LiveRatesCarriersList carrierIds={[]} />);
+
+		expect(wrapper.html()).toBeFalsy();
+	});
+
+	it('should not render when the list of accounts contains invalid carrier IDs', () => {
+		const wrapper = mount(<LiveRatesCarriersList carrierIds={['wc_services_fedex', 'wc_services_dhlecommerceasia']} />);
+
+		expect(wrapper.html()).toBeFalsy();
+	});
+
+	it('should render the USPS account information', () => {
+		const wrapper = mount(<LiveRatesCarriersList carrierIds={['wc_services_usps']} />);
+
+		expect(wrapper.contains(<div className="live-rates-carriers-list__carrier">USPS</div>)).toBe(true);
+		expect(wrapper.contains(<div className="live-rates-carriers-list__carrier">DHL Express</div>)).toBe(false);
+	});
+
+	it('should render the DHL Express account information', () => {
+		const wrapper = mount(<LiveRatesCarriersList carrierIds={['wc_services_dhlexpress']} />);
+
+		expect(wrapper.contains(<div className="live-rates-carriers-list__carrier">USPS</div>)).toBe(false);
+		expect(wrapper.contains(<div className="live-rates-carriers-list__carrier">DHL Express</div>)).toBe(true);
+	});
+});
```
:information_source: Missing spaces inside parens/braces.

Ferdev commented:
Tried to point out most of the missing spaces I found, but I may have missed some of them.

frosso commented:
@Ferdev thanks for pointing them out - I thought that the checks on the spacing would be automated? Running `npm run eslint` or `npm run eslint:fix` doesn't point out any issues, does it work for you?
It feels kinda cumbersome to have to manually add every single space 😄

Ferdev commented:
![](https://static.wikia.nocookie.net/memes-pedia/images/3/3b/300px-I-know-that-feel.jpg/revision/latest/scale-to-width-down/340?cb=20140215150923&path-prefix=es)

It doesn't work for me, either. I proposed ages ago about using https://github.com/Automattic/wp-prettier, which takes care of all the formatting for us. But we didn't ended up adopting it.

frosso commented:
@Ferdev I manually added the spaces to the files I added in this PR. I'm not sure if I followed the rules, since I don't see the rules that need to be enforced. I didn't change the files that I just modified.
I think that if the `JS Tests / JS Eslint and Test (push)` step passes, this should be good.

Ferdev commented:
Perfect, thanks @frosso!!

[Here](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/) you can find the JS coding standars. I just realized they recommend using JSHint instead. Maybe we should migrate eslint to JSHint to get rid of this problem

waclawjacek commented:
Regarding the multiline tooltp matter:

Here is what I did in #2314 to get the tooltip to wrap into multiple lines to make it match the design:

```css
.components-popover__content {
	white-space: normal;
	width: 200px;
}
```

And here is the result:

![Multiline tooltip in the "Shipping method" table explaining rate call usage over limit](https://user-images.githubusercontent.com/1759681/106327297-1d2fb800-627e-11eb-9d5e-e83dc40775ec.png)
